File: kickstart/2017/Practice-Round/A.java
import java.util.*;
import java.io.*;

public class Main {

    public static void main(String[] args) throws Exception {

        String path = "/Users/xushaopeng/Desktop/codejam/kickstart/2017/Practice-Round/";
        File file = new File(path + "A-large-practice.in");

        System.out.print(file.exists());
        Scanner in = new Scanner(file);
        //Scanner in = new Scanner(System.in);

        PrintWriter out = new PrintWriter(new FileWriter(path + "A-large-practice.out"));
        // PrintWriter out = new PrintWriter(System.out);

        int T = in.nextInt(), line = 1;

        while (--T >= 0) {
            int N = in.nextInt();
            String[] persons = new String[N];

            in.skip("\n");

            while (--N >= 0) {
                persons[N] = in.nextLine();
            }
            // 静态方法不能调用非静态方法，所以需要实例化 Main
            String ret = (new Main()).countryLeader(persons);

            out.println("Case #" + (line++) + ": " + ret);
        }

        in.close(); //关闭流并释放，这个很重要，否则是没有输出的
        out.close(); //关闭流并释放
    }
    private String countryLeader(String[] persons) {
        Map<String, Integer> myMap = new TreeMap();
        for (String s : persons) {
            Set<Character> tmp = new HashSet<Character>();
            for (char c : s.toCharArray()) {
                if (c != ' ') {
                    tmp.add(c);
                }
            }

            myMap.put(s, tmp.size());
        }
        String ret = "";
        int letters = 0;

        for (Map.Entry<String, Integer> entry : myMap.entrySet()) {
            if (entry.getValue() > letters) {
                letters = entry.getValue();
                ret = entry.getKey();
            }
        }
        return ret;
    }
}

File: kickstart/2017/Practice-Round/A.cpp
/**
    给定一组字符串，找到其中含有字母最多的一个字符串
    思路：用map对每个字符串的字母进行统计
*/
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <set>
#include <unordered_set>

using namespace std;
string countryLeader(vector<string> persons) {
    map<string, int> myMap;
    for (string s : persons) {
        unordered_set<char> tmp;
        for (char c : s) {
            if (c != ' ') tmp.insert(c);
        }
        myMap[s] = tmp.size();
    }
    string ret = "";
    int letters = 0;
    for (auto itor : myMap) {
        if (itor.second > letters) {
            letters = itor.second;
            ret = itor.first;
        }
    }
    return ret;
}
int main() {
    freopen("/Users/xushaopeng/Desktop/codejam/kickstart/2017/Practice-Round/A-large-practice.in", "r", stdin);
    freopen("/Users/xushaopeng/Desktop/codejam/kickstart/2017/Practice-Round/A-large-practice.out", "w", stdout);
    int T, line = 1;
    cin >> T;

    while(T--) {
        int N;
        cin >> N;

        vector<string> persons(N);
        // cin and getline skipping input
        // https://stackoverflow.com/questions/10553597/cin-and-getline-skipping-input
        // https://stackoverflow.com/questions/5739937/using-getlinecin-s-after-cin
        cin.ignore (numeric_limits<streamsize>::max(), '\n');

        while (N--) getline(cin, persons[N]);

        cout << "Case #" + to_string(line++) + ": " + countryLeader(persons) << endl;
    }

    return 0;
}
